lib/authorizeAccessToken.js

```diff
@@ -0,0 +1,187 @@
+ * decodes access token, verifies it, checks the decoded token claims against
+ * the provided scope, sets the `req.clientMetadata` property).
+ *
+ * @param {object} options - Options hashmap.
+ * @param {string} options.scope - Scope required for the route guarded by this
+ *   `authorize()` usage.
+ * @param {Function} [options.customValidate] - Optional custom authorization
+ *   callback with the signature `async ({req, claims}) => {}`.
+ * @param {object} [options.cache] - Optional LRU cache to store the decoded
+ *   token claims). The cache API must be compatible with the lru-cache NPM
+ *   module.
+ * @param {string[]} options.validIssuers - List of allowed issuers of
+ *   access tokens (e.g. 'https://admin.sandbox.nacsav.com').
+ * @param {Function} options.loadClientRegistration -
+ * @returns {Function} Returns an Express.js middleware route handler.
+ */
+export function authorizeAccessToken({
+  scope, cache, validIssuers, loadClientRegistration, customValidate}) {
+  return async (req, res, next) => {
+    try {
+      // Extract access token from Authorization: header (throw otherwise)
+      let clientAccessToken;
+      try {
+        // parse throws if the authorization header is missing
+        const auth = authorization.parse(req.get('authorization'));
+        if(!(auth.scheme === 'Bearer' && auth.token)) {
+          throw new Error('NotAllowedError');
+        }
+        ({token: clientAccessToken} = auth);
+      } catch(e) {
+        throw new Error('Missing or invalid "authorization" header.');
+      }
+
+      // validate and unpack the claims from the clientAccessToken
+      // use the cache as appropriate
+      // verify JWT w/cache-backed helper
+      const {valid, claims: tokenClaims, errors} = await _verifyClaimsFromJwt({
+        claimsCache: cache,
+        jwt: clientAccessToken,
+        validIssuers
+      });
+      if(!valid) {
+        const tokenError = new BedrockError(
+          'Invalid access token.', 'NotAllowedError', {
+            // errors is an array of errors which cannot be used as the
+            // BedrockError cause
+            errors,
+            httpStatusCode: 401,
+            public: true,
+          });
+        tokenError.errorId = 'invalid_token';
+        throw tokenError;
+      }
+
+      // const tokenClaims = {
+      //   iss: 'https://example.issuer.com',
+      //   client_id: 'myOauth2Client',
+      //   scope: 'read write something-else a-third-scope'
+      // };
+
+      // use the client_id claim to load the OAuth2 Client Registration
+      // Load Agent metadata based on client id
+      const {client_id} = tokenClaims;
+      let clientRegistration;
+      try {
+        (clientRegistration = await loadClientRegistration({client_id}));
+      } catch(e) {
+        throw new BedrockError(
+          'Could not load a registered agent for the provided client ID.',
+          'NotAllowedError', {
+            client_id,
+            httpStatusCode: 400,
```

Comment:
`NotAllowedError` should have consistent status code throughout.  403 is typical for this, but I think 401 might be OK as well.